[New KB Article] Understand rule severities (#1084)
returntocorp/semgrep-docs: docs/kb/rules/understand-severities.md

---
description: Understand how rule severity is determined.
tags:
  - Rules
  - Semgrep Registry
---

# How does Semgrep assign severity levels to rules?

## Semgrep Code

> Semgrep Code rules have one of three severity levels: ERROR (High), WARNING (Medium), or INFO (Low). The severity indicates how critical the issues are that a rule potentially detects.

Author: I don't object to this change, but since they're already in all caps I'm not sure it's necessary?

> The severity of a given rule is assigned by the rule author. For custom and third-party rules, the rule author's assignment is the source of truth.

Reviewer: ```suggestion
The severity of a given rule is assigned by the rule author. For custom and third-party rules, the rule author is the source of truth for the severity.
```

Reviewer: I was initially a little confused by the wording, does this sound any better?

Author: Huh, I thought it made sense - the severity is assigned by the author, and that assignment is the source of truth. Though I probably shouldn't be writing in passive voice anyway!

What about:

> The rule author assigns the rule severity. For custom and third-party rules, their severity assignment is the source of truth.

> As a best practice, severity for Semgrep Registry rules in the `security` category should be assigned by evaluating the combination of [likelihood](/docs/contributing/contributing-to-semgrep-rules-repository/#likelihood) and [impact](/docs/contributing/contributing-to-semgrep-rules-repository/#impact). 

Reviewer: I feel like this instills some doubt in our severity assigning process. Why do we say "as a best practice" and "should be". Would it be accurate to just say:

> As a best practice, severity for Semgrep Registry rules in the `security` category should be assigned by evaluating the combination of [likelihood](/docs/contributing/contributing-to-semgrep-rules-repository/#likelihood) and [impact](/docs/contributing/contributing-to-semgrep-rules-repository/#impact). 

Reviewer: ```suggestion
As a best practice, severity for Semgrep Registry rules in the `security` category is assigned by evaluating the combination of [likelihood](/docs/contributing/contributing-to-semgrep-rules-repository/#likelihood) and [impact](/docs/contributing/contributing-to-semgrep-rules-repository/#impact). 
```

Author: Currently we're not super consistent about it, so "is" might be a little strong. One day, I hope. :)

## Semgrep Supply Chain 

> Semgrep Supply Chain rules have one of four severity levels: Critical, High, Medium or Low. The score assigned to the CVE using the Common Vulnerability Scoring System (CVSS) score, or the severity value set by the GitHub Advisory Database, determines the severity in Semgrep Supply Chain. If a vulnerability is given a 9.0 or higher it is assigned CRITICAL.

Reviewer 2: Good article. I would add a link to the CVSS scoring: https://nvd.nist.gov/vuln-metrics/cvss

Author: Good call! I'll do that.

> Semgrep Supply Chain rules have one of four severity levels: Critical, High, Medium or Low. The score assigned to the CVE using the [Common Vulnerability Scoring System (CVSS) score])(https://nvd.nist.gov/vuln-metrics/cvss), or the severity value set by the GitHub Advisory Database, determines the severity in Semgrep Supply Chain. If a vulnerability is given a 9.0 or higher it is assigned CRITICAL.

Reviewer: I feel like we could either go without defining what Critical's score range is, or if we do, do it for all the other severities as well. Also I think "CRITICAL" should just be "Critical", like the first sentence.

Author: yeah you're right, the severity writing should be consistent.

I'll see if I can find out more about the other severities before we merge this :) I only got the 9+ score info from what we had internally.

Author: I was able to verify that when providing that example we were also referencing the same scoring rubric shown in the linked article, that just happened to be the level called out in the comment. I changed the wording to indicate that it's an example.
